src/realm/obj.cpp

```diff
@@ -886,67 +1008,101 @@ void Obj::to_json(std::ostream& out, size_t link_depth, std::map<std::string, st
+                        out_mixed_xjson(out, obj_link.get_obj_key());
+                        out << "}}";
+                    }
+                    else if (link_depth == 0 || link_depth == not_found) {
                         out << "{\"table\": \"" << target_table->get_name()
                             << "\", \"key\": " << obj_link.get_obj_key().value << "}";
                     }
                     else {
                         auto obj = target_table->get_object(obj_link.get_obj_key());
                         size_t new_depth = link_depth == not_found ? not_found : link_depth - 1;
-                        obj.to_json(out, new_depth, renames, followed);
+                        obj.to_json(out, new_depth, renames, followed, output_mode);
                     }
                 }
                 else {
-                    out_mixed(out, it.second);
+                    out_mixed(out, it.second, output_mode);
                 }
             }
             out << "}";
+            if (output_mode == output_mode_xjson) {
+                out << "}";
+            }
         }
         else {
             if (type == col_type_Link) {
                 TableRef target_table = get_target_table(ck);
                 auto k = get<ObjKey>(ck);
                 if (k) {
                     auto obj = get_linked_object(ck);
-                    if (!target_table->is_embedded() &&
-                        ((link_depth == 0) || (link_depth == not_found &&
-                                               std::find(followed.begin(), followed.end(), ck) != followed.end()))) {
+                    auto primary_key_coll = target_table->get_primary_key_column();
+
+                    if (output_mode == output_mode_xjson && !target_table->is_embedded() && primary_key_coll) {
+                        out << "{ \"$link\": { \"table\": \"" << get_target_table(ck)->get_name() << "\", \"key\": ";
+                        out_mixed_xjson(out, obj.get_any(primary_key_coll));
+                        out << "}}";
+                    }
+                    else if (!target_table->is_embedded() &&
+                             ((link_depth == 0) ||
+                              (link_depth == not_found &&
+                               std::find(followed.begin(), followed.end(), ck) != followed.end()))) {
                         out << "{\"table\": \"" << get_target_table(ck)->get_name()
                             << "\", \"key\": " << obj.get_key().value << "}";
                     }
                     else {
                         followed.push_back(ck);
                         size_t new_depth = link_depth == not_found ? not_found : link_depth - 1;
-                        obj.to_json(out, new_depth, renames, followed);
+
+                        if (output_mode == output_mode_xjson) {
+                            out << "{ \"$embeddedObj\": { \"table\": \"" << get_target_table(ck)->get_name()
```

Comment:
Why do we need a special $embeddedObj type annotation here? Embedded objects are already pretty natural in JSON. Is that to distinguish from schema-less dictionaries?